### PR TITLE
Apiserver supports proxy protocol

### LINKED ISSUE #76928: proxy protocol support in kube-apiserver
**What would you like to be added**:
Proxy Protocol support for kube-apiserver

**Why is this needed**:
When inspecting audit logs I've noticed load balancer private address is passed to the API - this makes bit harder to inspect API usage. Currently when enabling audit logs and enabling proxy-protocol on AWS ELB results in:
```
➜  ~ k cluster-info                                                                                                                                                

To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
Unable to connect to the server: remote error: tls: record overflow
```

I'd like to see real-user-IP address in audit logs.

### PR DESCRIPTION
**Pull request by joonghyunlee**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR allows kube-apiserver to recognize the proxy protocol if it can be worked behind HAProxy. Since the OpenStack magnum uses a Octavia load balancer which is based on HAProxy actually, users couldn't know the clients' IP via audit log.

**Which issue(s) this PR fixes**:
Fixes #76928

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
add supports for proxy protocol in kube-apiserver
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/options/serving.go [hunk 1]
```diff
@@ -340,11 +341,13 @@ func CreateListener(network, addr string, config net.ListenConfig) (net.Listener
 		network = "tcp"
 	}
 
-	ln, err := config.Listen(context.TODO(), network, addr)
+	l, err := config.Listen(context.TODO(), network, addr)
 	if err != nil {
 		return nil, 0, fmt.Errorf("failed to listen on %v: %v", addr, err)
 	}
 
+	ln := &proxyproto.Listener{Listener: l}
```
Comment (cheftako) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving.go:349]: What if someone doesn't want/need the proxy. If there is a problem with this code it cannot just be turned off with a flag.

Comment (joonghyunlee) [reply staging/src/k8s.io/apiserver/pkg/server/options/serving.go:349]: Thank you. I will add an option to enable/disable this feature.

File: go.mod [hunk 2]
```diff
@@ -181,6 +181,7 @@ replace (
 	github.com/andreyvit/diff => github.com/andreyvit/diff v0.0.0-20170406064948-c7f18ee00883
 	github.com/armon/circbuf => github.com/armon/circbuf v0.0.0-20150827004946-bbbad097214e
 	github.com/armon/go-metrics => github.com/armon/go-metrics v0.0.0-20180917152333-f0300d1749da
+	github.com/armon/go-proxyproto => github.com/armon/go-proxyproto v0.0.0-20200108142055-f0b8253b1507
```
Comment (adammw) [inline go.mod:184]: https://github.com/pires/go-proxyproto seems to be better maintained and support PROXY v2 (binary) format, any reason to stick with the armon version?

Comment (joonghyunlee) [reply go.mod]: I changed the library with pires's.

**Comment by oz123** [comment]
I was able to build and test these suggested changes, and it works fine. Any chance of merging this soon?

**Comment by BenTheElder** [comment]
No. https://www.kubernetes.dev/resources/release/#tldr

**Review by BenTheElder — COMMENTED** [review]
/approve
for `vendor/` changes
library looks fine

**Comment by oz123** [comment]
> No. https://www.kubernetes.dev/resources/release/#tldr

Thanks for the quick reply 😊 ok. I am aware of the release schedule   let me rephrase, merge to master :-) which you already approved. Thanks 👍

**Comment by aojea** [comment]
I agree with @mikedanese about this https://github.com/kubernetes/kubernetes/issues/76928#issuecomment-573278361

**Comment by oz123** [comment]
> I agree with @mikedanese about this [#76928 (comment)](https://github.com/kubernetes/kubernetes/issues/76928#issuecomment-573278361)
> 
> Given the availability of alternatives, I am not convinced the benefit is worth increasing the surface area of the kube-apiserver at the lowest layer.

Well, GCP and AWS are not for everyone. And as far as I understand even `X-Forward-for` is not for everyone (since it does not work with HTTPS). For the use case I mentioned OpenStack loadbalancers, it's very easy to add `X-Forward-for`, however, not all OpenStack clouds have TLS termination setup. Hence, one is forced to use `PROXY` (The openstack load balancer just forwards TCP to the API server which serves HTTPS).

IMO, the patch is so minor and the functionality is optional so that it's worth adding (but I am of course biased).

**Comment by BenTheElder** [comment]
To be a bit clearer:
- This won't be merging soon without agreement from the owners AND because we won't be merging anything but critical bug fixes for weeks assuming the release ships on time.
- I am *only* approving the dependency changes per our dependency review guidelines, *not* the change to api-server. Other people will review the rest of the changes.

**Comment by aojea** [comment]
> Well, GCP and AWS are not for everyone. 

I don't think that the comment from Mike means that, is just providing examples that implementing a frontend loadbalancer that keeps the source ip is technically possible, and there is more than one solution

> Since the OpenStack magnum uses a Octavia load balancer which is based on HAProxy actually, users couldn't know the clients' IP via audit log.

Is the solution to modify the apiserver to overcome that limitation of the Openstack project or should the Openstack project implement a loadbalancer solution that preserves the source ip? 

> For the use case I mentioned OpenStack loadbalancers, it's very easy to add X-Forward-for, however, not all OpenStack clouds have TLS termination setup. Hence, one is forced to use PROXY (The openstack load balancer just forwards TCP to the API server which serves HTTPS).

is really PROXY the only solution?  or is the only solution because of the implementation details of the loadbalancers in Opensack?

> IMO, the patch is so minor and the functionality is optional so that it's worth adding (but I am of course biased).

It is an apiserver OWNERS decision at the end,  but what happen if project foo with protocol bar has the same problem? do we add a new flag? this is also vendoring a new dependency, and that has always a cost ...

**Comment by oz123** [comment]
@aojea thanks for your reply. I updated my original response. However, I will answer shortly here too:

> is really PROXY the only solution? or is the only solution because of the implementation details of the loadbalancers in Opensack?

It's the only solution for TCP loadbalancers in front of an HTTPS server to preserve the original IP address.
It is not only and implementation detail of OpenStack. As far as I understand other cloud load balancers support it too (e.g AWS ELB see link above).

**Comment by fnkr** [comment]
/remove-lifecycle stale

 @joonghyunlee Can you please rebase this PR so it can be merged?

**Comment by dims** [comment]
/assign @liggitt

NOTE: Jordan this is a `size/L`  ...

/ok-to-test

File: staging/src/k8s.io/apiserver/pkg/server/options/serving.go [hunk 3]
```diff
@@ -244,6 +251,10 @@ func (s *SecureServingOptions) ApplyTo(config **server.SecureServingInfo) error
 		if err != nil {
 			return fmt.Errorf("failed to create listener: %v", err)
 		}
+
+		if s.ProxyProtocol {
+			s.Listener = &proxyproto.Listener{Listener: s.Listener}
+		}
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving.go:257]: what behavior does this change on established connections?

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/server/options/serving.go:257]: is there a reason this is only done in the `if s.Listener == nil {` branch? if there's already a listener, and s.ProxyProtocol is set, shouldn't this still be wrapped?

File: vendor/github.com/pires/go-proxyproto/LICENSE [hunk 4]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (liggitt) [inline vendor/github.com/pires/go-proxyproto/LICENSE:1]: this is licensed under Apache, but `Conn#readHeader()` indicates it is using code from github.com/armon/go-proxyproto which is MIT-licensed... I'm unsure about that

File: staging/src/k8s.io/apiserver/pkg/server/options/serving.go [hunk 5]
```diff
@@ -75,6 +76,9 @@ type SecureServingOptions struct {
 
 	// PermitAddressSharing controls if SO_REUSEADDR is used when binding the port.
 	PermitAddressSharing bool
+
+	// ProxyProtocol enables kube-apiserver to know the clients' IP via proxy protocol.
+        ProxyProtocol bool
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving.go:81]: is this tested anywhere?

**Comment by lavalamp** [comment]
I removed some not relevant SIG labels. Like I said in the issue, IMO SIG Auth needs to decide whether this is beneficial or risky.

**Comment by enj** [comment]
Reviewed on SIG Auth triage meeting on 2022-04-25.

This is a large change that warrants a KEP to understand the need and answer the various PRR questions, along with testing, feature flags, etc.  Without that, this cannot safely progress.

Before spending the time writing a KEP, I would recommend adding a discussion item to the [SIG Auth meeting agenda](https://docs.google.com/document/d/1woLGRoONE3EBVx-wTb4pvp4CI7tmLZ6lS26VTbosLKM) so we can have a high bandwidth conversation to make sure this functionality has enough consensus.

/hold
